src/kakarot/state.cairo

```diff
@@ -0,0 +1,518 @@
+                accounts_start=self.accounts_start,
+                accounts=accounts,
+                events_len=self.events_len,
+                events=self.events,
+                balances_start=self.balances_start,
+                balances=self.balances,
+                transfers_len=self.transfers_len,
+                transfers=self.transfers,
+            );
+            return (state, account);
+        } else {
+            // Otherwise read values from contract storage
+            local accounts: DictAccess* = accounts;
+            let (bytecode_len, bytecode) = Accounts.get_bytecode(address.evm);
+            // we assume that if there is no bytecode this is an EOA.
+            // in this context, the nonce is managed by Starkware and not accessible from within
+            // the contract, hence we put 0.
+            // It shouldn't have any impact
+            if (bytecode_len == 0) {
+                let account = new_account(code_len=bytecode_len, code=bytecode, nonce=0);
+                dict_write{dict_ptr=accounts}(key=address.evm, new_value=cast(account, felt));
+                tempvar state = new model.State(
+                    accounts_start=self.accounts_start,
+                    accounts=accounts,
+                    events_len=self.events_len,
+                    events=self.events,
+                    balances_start=self.balances_start,
+                    balances=self.balances,
+                    transfers_len=self.transfers_len,
+                    transfers=self.transfers,
+                );
+                return (state, account);
+            }
+
+            let (nonce) = IContractAccount.get_nonce(contract_address=address.starknet);
+            let account = new_account(code_len=bytecode_len, code=bytecode, nonce=nonce);
+            dict_write{dict_ptr=accounts}(key=address.evm, new_value=cast(account, felt));
+            tempvar state = new model.State(
+                accounts_start=self.accounts_start,
+                accounts=accounts,
+                events_len=self.events_len,
+                events=self.events,
+                balances_start=self.balances_start,
+                balances=self.balances,
+                transfers_len=self.transfers_len,
+                transfers=self.transfers,
+            );
+            return (state, account);
+        }
+    }
+
+    // @notice Create a new account
+    // @dev New accounts start at nonce=1.
+    // @param code_len The length of the code
+    // @param code The pointer to the code
+    // @param nonce The initial nonce
+    // @return The updated state
+    // @return The account
+    func new_account(code_len: felt, code: felt*, nonce: felt) -> model.Account* {
+        let (storage_start) = default_dict_new(0);
+        return new model.Account(
+            code_len=code_len,
+            code=code,
+            storage_start=storage_start,
+            storage=storage_start,
+            nonce=nonce,
```

Comment:
We should force the nonce to be 1 for contract account and 0 for eoa, maybe split into `new_contract_account` and `new_eoa`.

Reply:
I think EOA cannot be created as part of the EVM flow

Reply:
i don't understand

Reply:
`new_account` isn't used to create a new account. It's used to add an account in the cached state.

Reply:
but EOAs dont have storage no? they don't have any of the keys mentioned, they don't even have a `nonce` no? as far as the evm is concerned

Reply:
Reading the PR again I see that I was wrong on how the nonce is passed, so never mind.

Reply:
We need to add a support interface for EOA/CA so that when we touch we know if it's one or another. That being said, it has no impact as the nonce of a touched account with empty bytecode cannot have any impact. See my comment above

            // we assume that if there is no bytecode this is an EOA.
            // in this context, the nonce is managed by Starkware and not accessible from within
            // the contract, hence we put 0.
            // It shouldn't have any impact